src/Common/tests/System/Net/Http/LoopbackServer.cs

```diff
@@ -115,6 +130,247 @@ public static Task<List<string>> ReadRequestAndSendResponseAsync(Socket server,
             return lines;
         }
 
+        public static async Task<List<string>> ValidateAuthenticationAsync(Socket s, StreamReader reader, StreamWriter writer, string response, Options options)
+        {
+            // Send unauthorized response from server.
+            await ReadWriteAcceptedAsync(s, reader, writer, response);
+
+            // Read the request method.
+            string line = await reader.ReadLineAsync().ConfigureAwait(false);
+            int index = line != null ? line.IndexOf(' ') : -1;
+            string requestMethod = null;
+            if (index != -1)
+            {
+                requestMethod = line.Substring(0, index);
+            }
+
+            // Read the authorization header from client.
+            AuthenticationProtocols protocol = AuthenticationProtocols.None;
+            string clientResponse = null;
+            while (!string.IsNullOrEmpty(line = await reader.ReadLineAsync().ConfigureAwait(false)))
```

Comment:
It's ok for now, but related to @geoffkizer's refactoring, we may just want to have a method "ReadHeaders" that returns the list of headers sent by the client.  Then code like this can just process the array or dictionary of headers rather than interleaving request processing with test logic.

Reply:
And now that I'm reviewing Geoff's PR, I see exactly that:
https://github.com/dotnet/corefx/pull/27088/files#diff-fd50cecf004b510497450abfa7956a09R189